src/torrent.cpp

```diff
@@ -5664,99 +5675,66 @@ namespace {
 	void torrent::on_file_priority(disk_io_job const* j)
 	{
 		dec_refcount("file_priority");
+
 		boost::scoped_ptr<std::vector<boost::uint8_t> > p(j->buffer.priorities);
-		if (m_file_priority == *p) return;
 
-		// in this case, some file priorities failed to get set
-		m_file_priority = *p;
-		update_piece_priorities();
+		if (j->ret == piece_manager::no_error)
+		{
+			TORRENT_ASSERT(!j->error);
+			m_file_priority = *p;
+			update_piece_priorities();
+		}
+		else
+		{
+			// in this case, some file priorities failed to get set
+			TORRENT_ASSERT(j->error);
+			TORRENT_ASSERT(!m_share_mode);
 
-		if (alerts().should_post<file_error_alert>())
-			alerts().emplace_alert<file_error_alert>(j->error.ec
-				, resolve_filename(j->error.file), j->error.operation_str(), get_handle());
+			if (alerts().should_post<file_error_alert>())
+				alerts().emplace_alert<file_error_alert>(j->error.ec
+					, resolve_filename(j->error.file), j->error.operation_str(), get_handle());
 
-		set_error(j->error.ec, j->error.file);
-		pause();
+			set_error(j->error.ec, j->error.file);
+			pause();
+		}
 	}
 
 	void torrent::prioritize_files(std::vector<int> const& files)
 	{
 		INVARIANT_CHECK;
 
-		// this call is only valid on torrents with metadata
-		if (!valid_metadata() || is_seed()) return;
-
-		// the vector need to have exactly one element for every file
-		// in the torrent
-		TORRENT_ASSERT(int(files.size()) == m_torrent_file->num_files());
-
-		int limit = int(files.size());
-		if (valid_metadata() && limit > m_torrent_file->num_files())
-			limit = m_torrent_file->num_files();
-
-		if (int(m_file_priority.size()) < limit)
-			m_file_priority.resize(limit, 4);
-
-		std::copy(files.begin(), files.begin() + limit, m_file_priority.begin());
-
-		if (valid_metadata() && m_torrent_file->num_files() > int(m_file_priority.size()))
-			m_file_priority.resize(m_torrent_file->num_files(), 4);
-
-		// initialize pad files to priority 0
-		file_storage const& fs = m_torrent_file->files();
-		for (int i = 0; i < (std::min)(fs.num_files(), limit); ++i)
+		// store file priorities until we get the metadata
+		if (!valid_metadata())
 		{
-			if (!fs.pad_file_at(i)) continue;
-			m_file_priority[i] = 0;
+			m_file_priority.assign(files.begin(), files.end());
+			return;
 		}
 
+		std::vector<boost::uint8_t> const new_priority = fix_priorities(files, m_torrent_file->files());
+
+		if (!m_torrent_file->num_pieces() || is_seed() || new_priority == m_file_priority) { return; }
```

Comment:
it seems reasonable to move the checks that are independent of `new_priority` before doing the work of `fix_priorites()`, no?

Reply:
I believe that most compilers will optimize this code, i.e. `fix_priorities` is about to execute only when `!m_torrent_file->num_pieces() || is_seed()` checks are passed.

However I don't mind moving those checks up.

Reply:
only if `fix_priorities()` and `torrent_file::files()` can be proven to be pure.